exporter/splunkhecexporter/heartbeat.go

```diff
@@ -0,0 +1,154 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package splunkhecexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter"
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type heartbeater struct {
+	config     *Config
+	pushLogFn  func(ctx context.Context, ld plog.Logs) error
+	hbRunOnce  sync.Once
+	hbDoneChan chan struct{}
+
+	// Observability
+	heartbeatSuccessTotal *stats.Int64Measure
+	heartbeatErrorTotal   *stats.Int64Measure
+	tagMutators           []tag.Mutator
+}
+
+func newHeartbeater(config *Config, pushLogFn func(ctx context.Context, ld plog.Logs) error) *heartbeater {
+	prefix := config.HecHeartbeat.ObservabilityName
+	dimensions := config.HecHeartbeat.ObservabilityDimensions
+	tags := []tag.Key{}
+	tagMutators := []tag.Mutator{}
+	for key, val := range dimensions {
+		newTag := tag.MustNewKey(key)
+		tags = append(tags, newTag)
+		tagMutators = append(tagMutators, tag.Insert(newTag, val))
+	}
+
+	heartbeatSuccessTotal := stats.Int64(
+		fmt.Sprintf("%s_success_total", prefix),
+		"number of heartbeats made to the destination",
+		stats.UnitDimensionless)
+
+	heartbeatSuccessTotalView := &view.View{
+		Name:        heartbeatSuccessTotal.Name(),
+		Description: heartbeatSuccessTotal.Description(),
+		TagKeys:     tags,
+		Measure:     heartbeatSuccessTotal,
+		Aggregation: view.Sum(),
+	}
+
+	heartbeatErrorTotal := stats.Int64(
+		fmt.Sprintf("%s_error_total", prefix),
+		"number of heartbeats made to destination that failed",
+		stats.UnitDimensionless)
+
+	heartbeatErrorTotalView := &view.View{
+		Name:        heartbeatErrorTotal.Name(),
+		Description: heartbeatErrorTotal.Description(),
+		TagKeys:     tags,
+		Measure:     heartbeatErrorTotal,
+		Aggregation: view.Sum(),
+	}
+
+	_ = view.Register(heartbeatSuccessTotalView, heartbeatErrorTotalView)
+
+	return &heartbeater{
+		config:                config,
+		pushLogFn:             pushLogFn,
+		hbDoneChan:            make(chan struct{}),
+		heartbeatSuccessTotal: heartbeatSuccessTotal,
+		heartbeatErrorTotal:   heartbeatErrorTotal,
+		tagMutators:           tagMutators,
+	}
+}
+
+func (h *heartbeater) shutdown() {
+	close(h.hbDoneChan)
+}
+
+func (h *heartbeater) initHeartbeat(buildInfo component.BuildInfo) {
+	interval := h.config.HecHeartbeat.Interval
+	if interval == 0 {
+		return
+	}
+
+	h.hbRunOnce.Do(func() {
```

Comment:
Can we do with a sync.Once lock here?

Reply:
Not sure if I follow, `hbRunOnce` is a sync.Once that makes the block to run once.

Reply:
There is no thread safety needed here, or am I missing something?

Reply:
Since `initHeartbeat` spawns a goroutine, it is much safer to guard it with a `sync.Once`. Since it can't control the caller's behavior, it can spawn a lot of goroutine and hammer downstream if `initHeartbeat` called more than once. Ideally this shouldn't happen but this is more a safety net.

Reply:
I also don't think it has any value. It'll be called once per `heartbeater` instance. But you already have only one call to `initHeartbeat` after the heartbeater instance is created. So it's a safety net only against running `initHeartbeat` more than once on the same instance, which seems unnecessary